src/compiler/program.ts

```diff
@@ -386,6 +393,117 @@ namespace ts {
+            return currentOptions.configFile.text === newOptions.configFile.text;
+        }
+
+        return true;
+
+        function sourceFileUpToDate(sourceFile: SourceFile): boolean {
+            return sourceFile &&
+                sourceFile.version === getSourceVersion(sourceFile.path) &&
+                !hasInvalidatedResolution(sourceFile.path);
+        }
+    }
+
+    function shouldProgramCreateNewSourceFiles(program: Program, newOptions: CompilerOptions) {
+        // If any of these options change, we cant reuse old source file even if version match
+        const oldOptions = program && program.getCompilerOptions();
+        return oldOptions &&
+            (oldOptions.target !== newOptions.target ||
+                oldOptions.module !== newOptions.module ||
+                oldOptions.moduleResolution !== newOptions.moduleResolution ||
+                oldOptions.noResolve !== newOptions.noResolve ||
+                oldOptions.jsx !== newOptions.jsx ||
+                oldOptions.allowJs !== newOptions.allowJs ||
+                oldOptions.disableSizeLimit !== newOptions.disableSizeLimit ||
+                oldOptions.baseUrl !== newOptions.baseUrl ||
+                !equalOwnProperties(oldOptions.paths, newOptions.paths));
+    }
+
+    /**
+     * Updates the existing missing file watches with the new set of missing files after new program is created
+     */
+    export function updateMissingFilePathsWatch(program: Program, existingMap: Map<FileWatcher>,
+        createMissingFileWatch: (missingFilePath: Path) => FileWatcher,
+        closeExistingMissingFilePathFileWatcher: (missingFilePath: Path, fileWatcher: FileWatcher) => void) {
+
+        const missingFilePaths = program.getMissingFilePaths();
+        const newMissingFilePathMap = arrayToSet(missingFilePaths);
+        // Update the missing file paths watcher
+        return mutateExistingMapWithNewSet(
+            existingMap, newMissingFilePathMap,
+            // Watch the missing files
+            createMissingFileWatch,
+            // Files that are no longer missing (e.g. because they are no longer required)
+            // should no longer be watched.
+            closeExistingMissingFilePathFileWatcher
+        );
+    }
+
+    export type WildCardDirectoryWatchers = { watcher: FileWatcher, recursive: boolean };
+
+    /**
+     * Updates the existing wild card directory watcyhes with the new set of wild card directories from the config file after new program is created
+     */
+    export function updateWatchingWildcardDirectories(existingWatchedForWildcards: Map<WildCardDirectoryWatchers>, wildcardDirectories: Map<WatchDirectoryFlags>,
+        watchDirectory: (directory: string, recursive: boolean) => FileWatcher,
+        closeDirectoryWatcher: (directory: string, watcher: FileWatcher, recursive: boolean, recursiveChanged: boolean) => void) {
+        return mutateExistingMap(

```

Comment:
This would be more readable if `mutateExistingMap` took an object, rather than many positional arguments.